### PR TITLE
MOBT-337: Handling partial periods in timezone mapping

### PR DESCRIPTION
**Pull request by bayliffe**
## What it does
We need to produce local-time zone daily summary weather symbols. These need to be updated throughout the day, so must make use of partial period data (i.e. a mode calculated over less than a full 24 hours). The time zone mapping code uses the time dimensions formed by merging cubes from different validity times together. However, with partial periods, several of the cubes may have the same starting lower bound, e.g. the 12Z-18Z cube and the 12Z-12Z cube. These cannot be combined into a dimension coordinate as the bounds are not monotonically increasing. This change manipulates the anonymous coordinate that is associated with the resulting time auxiliary coordinate to enable the time zone mapping to occur as before.

A further change is to explicitly extract the bounds for each location from the source data. Previously, as only complete periods were used, one bound range was calculated and applied relative to all of the extracted time points; this will give the incorrect bounds when using partial periods.

In addition this PR handles incomplete periods. These are differentiated from partial periods by being at the end of the forecast range. These don't update an ongoing period but rather report a result for an incomplete day which is not desirable. For example we might provide a day max temperature based on just the first 2 hours of the day. Note that the impact here is only on weather symbols at present as the other local time zone products do not allow partial or incomplete periods in their suite definitions. As the two cannot be differentiated at the suite level and we want partial period updating for WX this must now be handled.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

File: improver_tests/utilities/temporal/test_TimezoneExtraction.py [hunk 1]
```diff
@@ -170,22 +171,44 @@ def test_create_output_cube(with_cell_method):
...
         cube.remove_coord("time")
         with pytest.raises(
             ValueError, match=r"Expected coords on input_cube: time, y, x "
         ):
-            plugin.check_input_cube_dims(cube, timezone_cube)
+            plugin.check_input_cube_dims(cube)
+
+
+def test_check_aux_time_coord():
+    """Checks that check_input_cube_dims can work with a auxiliary time
```
Comment (MoseleyS) [inline improver_tests/utilities/temporal/test_TimezoneExtraction.py:191]:
```suggestion
    """Checks that check_input_cube_dims can work with an auxiliary time
```

File: improver_tests/utilities/temporal/test_TimezoneExtraction.py [hunk 2]
```diff
@@ -170,22 +171,44 @@ def test_create_output_cube(with_cell_method):
...
+    coordinate. This occurs when partial periods are allowed. In these
+    cases the lower time bound may be the same between different time
+    points, preventing the creation of a dimension coordinate that
+    required monotonically increasing bounds. In this case a anonymous
```
Comment (MoseleyS) [inline improver_tests/utilities/temporal/test_TimezoneExtraction.py:195]:
```suggestion
    requires monotonically increasing bounds. In this case a anonymous
```

File: improver_tests/utilities/temporal/test_TimezoneExtraction.py [hunk 3]
```diff
@@ -170,22 +171,44 @@ def test_create_output_cube(with_cell_method):
...
+    reordering. This should not be present on the output, but the time
+    coordinate should have been moved to the last dimension."""
+    cube = make_input_cube([3, 4])
+    timezone_cube = make_timezone_cube()
+    iris.util.demote_dim_coord_to_aux_coord(cube, "time")
+
+    plugin = TimezoneExtraction()
+    plugin.timezone_cube = timezone_cube
+    plugin.check_input_cube_dims(cube)
+    print(cube)
```
Comment (MoseleyS) [inline improver_tests/utilities/temporal/test_TimezoneExtraction.py:209]: I think this print statement should be removed

File: improver_tests/utilities/temporal/test_TimezoneExtraction.py [hunk 4]
```diff
@@ -170,22 +171,44 @@ def test_create_output_cube(with_cell_method):
...
+    assert cube.coord_dims("time") == tuple([cube.ndim - 1])
+    assert "time_points" not in [crd.name() for crd in cube.coords()]
```
Comment (MoseleyS) [inline improver_tests/utilities/temporal/test_TimezoneExtraction.py:211]: Checking for a negative...

What we expect here is that we have exactly one dimension coordinate that is anonymous. How about
```
assert len(cube.dim_coords) == len(cube.shape) - 1
```
I'd rather check that the coordinate associated with the auxilliary `time` coord is the one that has no dimension coord, but I haven't figured that out yet.

File: improver_tests/utilities/temporal/test_TimezoneExtraction.py [hunk 5]
```diff
@@ -247,23 +279,99 @@ def test_process(with_percentiles, input_as_cube, input_has_time_bounds):
     local_time = datetime(2017, 11, 10, 5, 0)
     timezone_cube = make_timezone_cube()
     row1 = [cube.coord("time").units.date2num(datetime(2017, 11, 10, 4, 0))] * 4
-    row2 = [cube.coord("time").units.date2num(datetime(2017, 11, 10, 5, 0))] * 4
-    row3 = [cube.coord("time").units.date2num(datetime(2017, 11, 10, 6, 0))] * 4
-    expected_times = [row1, row2, row3]
+    row3 = [cube.coord("time").units.date2num(datetime(2017, 11, 10, 5, 0))] * 4
+    expected_times = [row1, row1, row3]
```
Comment (MoseleyS) [inline improver_tests/utilities/temporal/test_TimezoneExtraction.py:283]: I wish I could think of a better way of writing this.

**Review by MoseleyS — CHANGES_REQUESTED** [review]
This is a bit mind-bending, but I think I followed it.

File: improver/__init__.py [hunk 6]
```diff
@@ -84,7 +84,9 @@ def __call__(self, *args, **kwargs):
         from iris.cube import Cube
 
         result = super().__call__(*args, **kwargs)
-        if isinstance(result, Cube):
+        if result is None:
+            return result
+        elif isinstance(result, Cube):
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/__init__.py:89]: I've modified this and added a test. I don't think it will do anything daft.

File: improver/utilities/temporal.py [hunk 7]
```diff
@@ -489,30 +505,51 @@ def check_input_cube_time(self, input_cube: Cube, local_time: datetime) -> None:
...
+                # When mapping to timezones, the input cube contains data at times that
+                # correspond with different timezones. The earliest times correspond to
+                # the eastern most timezones, and the latest time to the western most.
+                # Incomplete periods in the east are indicative of a same day partial
+                # period, i.e. producing a 15-00 instead of 00-00 forecast for Japan.
+                # These are desirable as they update the same day forecast.
+                # Incomplete periods in the west are indicative of a data shortfall for
+                # representing the whole period at the longest lead-times. These should
+                # not return output as the daily summary would not represent a whole day
+                # for these regions.
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/utilities/temporal.py:530]: Well done for wrapping your mind around this stuff. You're right, I've reworded the comment to highlight that the issue is shorter bound periods in the west than in the east which is indicative if incomplete periods at the longer lead-times.

File: improver/utilities/temporal.py [hunk 8]
```diff
@@ -427,60 +427,76 @@ def _fill_timezones(self, input_cube: Cube) -> None:
...
-        enforce_coordinate_ordering(
-            self.timezone_cube, ["UTC_offset"], anchor_start=False
-        )
+            try:
+                time_aux_dim = input_cube.coord_dims(time_coord_name)
+            except CoordinateNotFoundError:
+                raise ValueError(
+                    f"Expected coords on input_cube: time, y, x ({expected_coords})."
+                    f"Found {cube_coords}"
+                )
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/utilities/temporal.py:465]: Both changes made. The ValueError existed before so I simply didn't modify the type; I have now.

File: improver/utilities/temporal.py [hunk 9]
```diff
@@ -427,60 +427,76 @@ def _fill_timezones(self, input_cube: Cube) -> None:
...
+            else:
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/utilities/temporal.py:466]: Removed.

File: improver/utilities/temporal.py [hunk 10]
```diff
@@ -427,60 +427,76 @@ def _fill_timezones(self, input_cube: Cube) -> None:
...
+                if time_aux_dim:
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/utilities/temporal.py:467]: Removed.

File: improver/utilities/temporal.py [hunk 11]
```diff
@@ -427,60 +427,76 @@ def _fill_timezones(self, input_cube: Cube) -> None:
...
+                    tp = input_cube.coord(time_coord_name).copy()
+                    tp.bounds = None
+                    tp.rename("time_points")
+                    input_cube.add_aux_coord(tp, 0)
+                    iris.util.promote_aux_coord_to_dim_coord(input_cube, "time_points")
+                    time_coord_name = "time_points"
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/utilities/temporal.py:473]: Done. I've not modified the internal name to include temporary, but I have renamed the variable to highlight this.

File: improver_tests/utilities/temporal/test_TimezoneExtraction.py (hunk 4 above)
Comment (bayliffe) [reply improver_tests/utilities/temporal/test_TimezoneExtraction.py:211]: I've added your suggested test to highlight that an anonymous dimension exists on the output. I've retained the explicit check that the "time_points" coord has been removed from the output as well just to demonstrate that this temporary coordinate won't escape the context of this function and head downstream.

File: improver_tests/test_PostProcessingPlugin.py [hunk 12]
```diff
@@ -108,3 +108,13 @@ def test_title_mandatory_attribute_default(dummy_plugin, plugin_input, is_list):
...
     assert_title_attribute(result, expected_title)
+
+
+def test_non_cubes(dummy_plugin):
+    """Test non-cube types are returned unchanged"""
+    result = dummy_plugin()(None)
+    assert result is None
+
+    iterable_input = ["list", "of", "strings"]
+    result = dummy_plugin()(iterable_input)
```
Comment (MoseleyS) [inline improver_tests/test_PostProcessingPlugin.py:119]: This test doesn't prevent the plugin modifying the input in place.
```suggestion
    result = dummy_plugin()(iterable_input.copy())
```

**Review by MoseleyS — CHANGES_REQUESTED** [review]
Just one more thing...

**Review by MoseleyS — DISMISSED** [review]

File: improver/utilities/temporal.py [hunk 13]
```diff
@@ -489,30 +504,52 @@ def check_input_cube_time(self, input_cube: Cube, local_time: datetime) -> None:
...
         self.timezone_cube = self.timezone_cube[:, :, ::-1]
+
         timezone_coord = self.timezone_cube.coord("UTC_offset")
         timezone_coord.convert_units("seconds")
         output_times = [
             local_time - timedelta(seconds=np.int(offset))
             for offset in timezone_coord.points
         ]
         if input_time_points != output_times:
+            if input_cube.coord("time").bounds is not None:
```
Comment (gavinevans) [inline improver/utilities/temporal.py:519]: How does the `if input_time_points != output_times` work for a partial period case? I would have expected the timezone cube to contain all possible UTC offsets, whilst the `input_time_points` would only contain a partial period, and therefore the `if input_time_points != output_times` check would be true, however, this doesn't seem to be how the `test_partial_period_update` test is set up given that `input_time_points == output_times`. Is this correct?

**Review by gavinevans — CHANGES_REQUESTED** [review]
Thanks @bayliffe 👍 

I've added some mostly minor comments.

File: improver/__init__.py [hunk 14]
```diff
@@ -86,9 +87,10 @@ def __call__(self, *args, **kwargs):
         result = super().__call__(*args, **kwargs)
         if isinstance(result, Cube):
             self.post_processed_title(result)
-        else:
-            for cube in result:
-                self.post_processed_title(cube)
+        elif isinstance(result, Iterable) and not isinstance(result, str):
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/__init__.py:90]: You're asking for the inverse of what Stephen requested. I originally explicitly handled cases of None being returned and nothing else, but Stephen suggested that anything that's not a cube just be returned untouched by this plugin. It was a reasonable suggestion as  this plugin is not meant to police what we return, only modify title strings if present. As such I went with Stephen's suggestion.

[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver/__init__.py:90]: Arm-wrestle?

File: improver/utilities/temporal.py (hunk 13 above)
Comment (bayliffe) [reply improver/utilities/temporal.py:519]: Your question, as discussed, gets to the heart of the limitations of these changes. At present the partial period update still requires that there be at least some valid data in every timezone. This means that eastern timezones can be updated all day down to a single remaining hour. However, western timezones will get far fewer updates through the first day of the forecast; once the east is in a different day we stop updating the west.

It is obviously required that all timezones can be updated throughout the first day to ensure consistency between the hourly symbols and the summary. As such another ticket has been created to go back and make further modifications to this code to enable this behaviour. This may require that masked data is returned for eastern timezones, or that the final hour for each is used to populate the summary once the day has passed in these timezones.

The fact that this behaviour doesn't yet exist is why this if statement is currently as it is.

File: improver/utilities/temporal.py [hunk 15]
```diff
@@ -549,9 +586,11 @@ def process(
         else:
             input_cube = MergeCubes()(CubeList(input_cubes))
 
-        self.check_timezones_are_unique(timezone_cube)
-        self.check_input_cube_dims(input_cube, timezone_cube)
-        self.check_input_cube_time(input_cube, local_time)
+        self.timezone_cube = timezone_cube.copy()
+        self.check_timezones_are_unique()
+        if not self.check_input_cube_time(input_cube, local_time):
+            return None
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/utilities/temporal.py:592]: Made Optional which implied None may be returned.

File: improver/utilities/temporal.py [hunk 16]
```diff
@@ -427,60 +427,75 @@ def _fill_timezones(self, input_cube: Cube) -> None:
...
-        time_coord = input_cube.coord("time")
-        point = time_coord.points[0]
-        if time_coord.has_bounds():
-            bounds = time_coord.bounds[0]
-            return bounds - point
-        else:
-            return None
-
-    def check_input_cube_dims(self, input_cube: Cube, timezone_cube: Cube) -> None:
+        if input_cube.coord("time").bounds is not None:
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/utilities/temporal.py:430]: Done.

File: improver/utilities/temporal.py [hunk 17]
```diff
@@ -427,60 +427,75 @@ def _fill_timezones(self, input_cube: Cube) -> None:
...
-        )
+            try:
+                time_aux_dim = input_cube.coord_dims(time_coord_name)
+            except CoordinateNotFoundError as err:
+                raise CoordinateNotFoundError(
+                    f"Expected coords on input_cube: time, y, x ({expected_coords})."
+                    f"Found {cube_coords}"
+                ) from err
+
+            temporary_time_dimcoord = input_cube.coord(time_coord_name).copy()
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/utilities/temporal.py:467]: That's not quite what is going on here, it's the construction of a new time coordinate when iris can't make one because of non-monotonic bounds. I've significantly extended the doc string to explain this.

File: improver/utilities/temporal.py [hunk 18]
```diff
@@ -427,60 +427,75 @@ def _fill_timezones(self, input_cube: Cube) -> None:
...
             raise ValueError(
                 "Spatial coordinates on input_cube and timezone_cube do not match."
             )
 
-    def check_input_cube_time(self, input_cube: Cube, local_time: datetime) -> None:
+    def check_input_cube_time(self, input_cube: Cube, local_time: datetime) -> bool:
         """Ensures input cube and timezone_cube cover exactly the right points and that
-        the time and UTC_offset coords have the same order.
+        the time and UTC_offset coords have the same order. If not a warning is raised
+        and the plugin will return nothing.
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/utilities/temporal.py:487]: I've tried to clarify this as it's not very clear. It will also have to be modified again when we allow empty periods.

File: improver/utilities/temporal.py [hunk 19]
```diff
@@ -549,9 +586,11 @@ def process(
         else:
             input_cube = MergeCubes()(CubeList(input_cubes))
 
-        self.check_timezones_are_unique(timezone_cube)
-        self.check_input_cube_dims(input_cube, timezone_cube)
-        self.check_input_cube_time(input_cube, local_time)
+        self.timezone_cube = timezone_cube.copy()
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/utilities/temporal.py:589]: I've added a load of extended documentation that hopefully covers this. I've also included the need to handle missing timezones as we discussed yesterday. This is of course not included in this PR so I've marked it as not implemented in to the docs until we get there.

File: improver_tests/utilities/temporal/test_TimezoneExtraction.py [hunk 20]
```diff
@@ -170,22 +172,44 @@ def test_create_output_cube(with_cell_method):
 @pytest.mark.parametrize("include_time_coord", (True, False))
 def test_check_input_cube_dims(include_time_coord):
     """Checks that check_input_cube_dims can differentiate between an input cube
-    with time, y, x coords and one where time is missing. Also checks that timezone_cube
-    has been reordered correctly."""
+    with time, y, x coords and one where time is missing."""
     cube = make_input_cube([3, 4])
     timezone_cube = make_timezone_cube()
     plugin = TimezoneExtraction()
+    plugin.timezone_cube = timezone_cube
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver_tests/utilities/temporal/test_TimezoneExtraction.py:179]: Changed everywhere.

**Review by MoseleyS — DISMISSED** [review]
Re-approved

File: improver/__init__.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (gavinevans) [reply improver/__init__.py:90]: OK, how about just adding a few unit tests, just to show that things other than `None` can be passed through? e.g.

```
    result = dummy_plugin()(0)
    assert result == 0

    result = dummy_plugin()("cat")
    assert result is "cat"
```

**Review by gavinevans — CHANGES_REQUESTED** [review]
Thanks @bayliffe 👍 

Thanks for the extended documentation. I've added a few very minor comments.

File: improver/__init__.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/__init__.py:90]: Added.

File: doc/source/extended_documentation/utilities/temporal/timezone_extraction.rst [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+    :scale: 100 %
+    :alt: UK area timezone masks.
+
+The plugin also takes a target local time as input, which is used to ensure
+that the correct forecasts are provided. The final input to the plugin is
+forecasts valid at times that span the UTC offset range required by the
+ancillary and which are valid at the target local times.
+
+The forecast data is stacked to align the time coordinate with the UTC offset
+coordinate in the ancillary and the mask is then applied. Following this step
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply doc/source/extended_documentation/utilities/temporal/timezone_extraction.rst:33]: I've added one for this bit. at least.

**Review by gavinevans — APPROVED** [review]
